From 3aaefb46601b0a5f56b560d5085bff4486dc440a Mon Sep 17 00:00:00 2001
From: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Date: Wed, 23 Jul 2014 13:27:03 -0400
Subject: [input] synaptics: report INPUT_PROP_TOPBUTTONPAD property

Message-id: <1406122030-5632-8-git-send-email-benjamin.tissoires@redhat.com>
Patchwork-id: 86349
O-Subject: [RHEL7.1 / RHEL7.0.z PATCH 07/14] Input: synaptics - report INPUT_PROP_TOPBUTTONPAD property
Bugzilla: 1093449
RH-Acked-by: Aristeu S. Rozanski F. <arozansk@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Peter Hutterer <peter.hutterer@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1093449
Upstream: since v3.15, backported in stable
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7722437
Test: tested by myself on a t440s and a t410, was extensively tested in RHEL 6.6

commit 43e19888b1fe2a3e8a5543030c5b286cde38b3f5
Author: Hans de Goede <hdegoede@redhat.com>
Date:   Sat Apr 19 22:26:41 2014 -0700

    Input: synaptics - report INPUT_PROP_TOPBUTTONPAD property

    Check PNP ID of the PS/2 AUX port and report INPUT_PROP_TOPBUTTONPAD
    property for for touchpads with top button areas.

    Signed-off-by: Hans de Goede <hdegoede@redhat.com>
    Signed-off-by: Dmitry Torokhov <dmitry.torokhov@gmail.com>

Signed-off-by: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/input/mouse/synaptics.c b/drivers/input/mouse/synaptics.c
index d8d49d1..9d75410 100644
--- a/drivers/input/mouse/synaptics.c
+++ b/drivers/input/mouse/synaptics.c
@@ -117,6 +117,44 @@ void synaptics_reset(struct psmouse *psmouse)
 }
 
 #ifdef CONFIG_MOUSE_PS2_SYNAPTICS
+/* This list has been kindly provided by Synaptics. */
+static const char * const topbuttonpad_pnp_ids[] = {
+ "LEN0017",
+ "LEN0018",
+ "LEN0019",
+ "LEN0023",
+ "LEN002A",
+ "LEN002B",
+ "LEN002C",
+ "LEN002D",
+ "LEN002E",
+ "LEN0033", /* Helix */
+ "LEN0034", /* T431s, T540, X1 Carbon 2nd */
+ "LEN0035", /* X240 */
+ "LEN0036", /* T440 */
+ "LEN0037",
+ "LEN0038",
+ "LEN0041",
+ "LEN0042", /* Yoga */
+ "LEN0045",
+ "LEN0046",
+ "LEN0047",
+ "LEN0048",
+ "LEN0049",
+ "LEN2000",
+ "LEN2001",
+ "LEN2002",
+ "LEN2003",
+ "LEN2004", /* L440 */
+ "LEN2005",
+ "LEN2006",
+ "LEN2007",
+ "LEN2008",
+ "LEN2009",
+ "LEN200A",
+ "LEN200B",
+ NULL
+};
 
 /*****************************************************************************
  * Synaptics communications functions
@@ -1255,8 +1293,10 @@ static void set_abs_position_params(struct input_dev *dev,
  input_abs_set_res(dev, y_code, priv->y_res);
 }
 
-static void set_input_params(struct input_dev *dev, struct synaptics_data *priv)
+static void set_input_params(struct psmouse *psmouse,
+        struct synaptics_data *priv)
 {
+ struct input_dev *dev = psmouse->dev;
  int i;
 
  /* Things that apply to both modes */
@@ -1325,6 +1365,17 @@ static void set_input_params(struct input_dev *dev, struct synaptics_data *priv)
 
  if (SYN_CAP_CLICKPAD(priv->ext_cap_0c)) {
   __set_bit(INPUT_PROP_BUTTONPAD, dev->propbit);
+  /* See if this buttonpad has a top button area */
+  if (!strncmp(psmouse->ps2dev.serio->firmware_id, "PNP:", 4)) {
+   for (i = 0; topbuttonpad_pnp_ids[i]; i++) {
+    if (strstr(psmouse->ps2dev.serio->firmware_id,
+        topbuttonpad_pnp_ids[i])) {
+     __set_bit(INPUT_PROP_TOPBUTTONPAD,
+        dev->propbit);
+     break;
+    }
+   }
+  }
   /* Clickpads report only left button */
   __clear_bit(BTN_RIGHT, dev->keybit);
   __clear_bit(BTN_MIDDLE, dev->keybit);
@@ -1593,7 +1644,7 @@ static int __synaptics_init(struct psmouse *psmouse, bool absolute_mode)
        priv->capabilities, priv->ext_cap, priv->ext_cap_0c,
        priv->board_id, priv->firmware_id);
 
- set_input_params(psmouse->dev, priv);
+ set_input_params(psmouse, priv);
 
  /*
   * Encode touchpad model so that it can be used to set
-- 
1.7.1